src/Validation/URLValidationRESTController.php

```diff
@@ -0,0 +1,278 @@
+<?php
+/**
+ * REST endpoint providing theme scan results.
+ *
+ * @package AMP
+ * @since 2.1
+ */
+
+namespace AmpProject\AmpWP\Validation;
+
+use AMP_Validated_URL_Post_Type;
+use AMP_Validation_Error_Taxonomy;
+use AmpProject\AmpWP\DevTools\UserAccess;
+use AmpProject\AmpWP\Infrastructure\Delayed;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+use WP_Error;
+use WP_REST_Controller;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * URLValidationRESTController class.
+ *
+ * @since 2.1
+ */
+final class URLValidationRESTController extends WP_REST_Controller implements Delayed, Service, Registerable {
+
+	/**
+	 * URLValidationProvider instance.
+	 *
+	 * @var URLValidationProvider
+	 */
+	private $url_validation_provider;
+
+	/**
+	 * UserAccess instance.
+	 *
+	 * @var UserAccess
+	 */
+	private $dev_tools_user_access;
+
+	/**
+	 * Response schema.
+	 *
+	 * @var array
+	 */
+	protected $schema;
+
+	/**
+	 * Get the action to use for registering the service.
+	 *
+	 * @return string Registration action to use.
+	 */
+	public static function get_registration_action() {
+		return 'rest_api_init';
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @param URLValidationProvider $url_validation_provider URLValidationProvider instance.
+	 * @param UserAccess            $dev_tools_user_access UserAccess instance.
+	 */
+	public function __construct( URLValidationProvider $url_validation_provider, UserAccess $dev_tools_user_access ) {
+		$this->namespace               = 'amp/v1';
+		$this->url_validation_provider = $url_validation_provider;
+		$this->dev_tools_user_access   = $dev_tools_user_access;
+	}
+
+	/**
+	 * Registers all routes for the controller.
+	 */
+	public function register() {
+		register_rest_route(
+			$this->namespace,
+			'/validate-post-url/(?P<id>[\d]+)',
+			[
+				'args'   => [
+					'id'            => [
+						'description' => __( 'Unique identifier for the object.', 'amp' ),
+						'required'    => true,
+						'type'        => 'integer',
+					],
+					'preview_id'    => [
+						'description' => __( 'Unique identifier for the preview.', 'amp' ),
+						'required'    => false,
+						'type'        => 'integer',
+					],
+					'preview_nonce' => [
+						'description' => __( 'Preview nonce string.', 'amp' ),
+						'required'    => false,
+						'type'        => 'string',
+					],
```

Comment:
I realized we would do well to validate these parameters. 

1. Ensure the `id` and `preview_id` are positive integers which refer to an actual post.
2. Ensure the `preview_nonce` is sanitized.

Reply:
This has been addressed in 669fdce.

Note that we don't have to manually check if `preview_id` is a positive integer since – as far as I understand and confirmed in tests – it's something done by the REST controller itself. This is why the only thing I've added is a `validate_callback` that checks if a post exists. With this, it seems the endpoint returns errors as expected:

**String:**
<img width="1007" alt="Screenshot 2021-02-26 at 14 54 42" src="https://user-images.githubusercontent.com/478735/109308661-9a7d2700-7842-11eb-85e4-62071dee0d0f.png">

**Negative number:**
<img width="1015" alt="Screenshot 2021-02-26 at 14 16 46" src="https://user-images.githubusercontent.com/478735/109308469-5722b880-7842-11eb-9820-7b87c41620cb.png">

**Inexistent post:**
<img width="1010" alt="Screenshot 2021-02-26 at 14 17 04" src="https://user-images.githubusercontent.com/478735/109308507-67d32e80-7842-11eb-9370-296a83ca60f8.png">

Also, the `preview_nonce` is now verified in a similar way as Core does in [`_show_post_preview()`](https://developer.wordpress.org/reference/functions/_show_post_preview/)

<img width="976" alt="Screenshot 2021-02-26 at 14 48 42" src="https://user-images.githubusercontent.com/478735/109308923-ec25b180-7842-11eb-95e2-a909fd0d211c.png">

<img width="750" alt="Screenshot 2021-02-26 at 14 58 04" src="https://user-images.githubusercontent.com/478735/109309041-0e1f3400-7843-11eb-8789-80c825dcb8b9.png">

Reply:
While working on this I realized there's no error handling in our UI (editor sidebar).

I'd say it's not mission-critical, but something we should definitely look into shortly. How about adding error reporting in the follow-up PR #5929 I've already started?

@jwold Having your input on this would be much appreciated.

Reply:
Yes, showing the any error resulting from validation can be incorporated into #5929.

Reply:
Agreed. Will followup on #5929